File: leetcode/medium/700-search-bst/iterative.py
"""
  @author Vortexx2
  Problem 700 - Search in a BST

  Runtime - 76 ms
  Memory Usage - 16 MB
"""
from typing import List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def searchBST(self, root: TreeNode, val: int) -> TreeNode:

    if root.val == val:
      return root

    while (root is not None):

      if root.val == val:
        return root

      if (root.val < val):
        root = root.right

      else:
        root = root.left

    return None

File: leetcode/medium/700-search-bst/iterative.cpp
/**
 * @file iterative.cpp
 * @author Vortexx2
 * @brief Problem 700 - Search in a BST
 * @date 25-07-2021
 * 
 * Runtime - 28 ms
 * Memory Usage - 34.7 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
#include <list>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  TreeNode *searchBST(TreeNode *root, int val)
  {
    while (root)
    {
      if (root->val == val)
        return root;

      if (root->val < val)
        root = root->right;

      else
        root = root->left;
    }

    return NULL;
  }
};